### ISSUE TITLE
[question] Versioning of conan packages for pre-built libraries

### ISSUE THREAD

**Issue by fischjo3SICKAG** [conan-io/conan#7415]
What is the 'conanian way' to solve the following situation?
Let X be a third party library with version A.B.C that shall be packaged as conan package using a recipe with version A.B.C. After successfully packaging it, recipe and package is uploaded to repository as released. Then, another software Y consumes the released recipe and becomes released too.
After a few months, a new software Z is being developed. Z consumes X, version A.B.C, too. We find out that recipe of X has to be changed (e.g. due to a bug). How can we fix and release X without incrementing its version and without changing the dependency of Y?

- [x ] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @jfischer1 

This is a very good question. There are 2 modes of operation in Conan right now:

- Without revisions: the default mode. A package can be regenerated with exactly the same version, it will be overwritten in the local cache, and overwritten in the server when it is uploaded (if the server has that permissions, typically it is possible unless explicitly configured for not allowing overwrites).
  With this mode, the binaries that are not re-built using the modified recipe (conanfile.py and/or sources), become "outdated".
  It is possible for other consumers, to install the updated recipes and binaries providing the ``--update`` argument at install time.  
  By default, if what is in the local cache satisfies the versions, that is used, and not automatically updated at ``conan install`` unless ``--update`` is used

- With revisions: needs to be enabled in conan.conf. Will be the default and only mode in Conan 2.0. With this mode, when you do a modification of a recipe (conanfile and/or sources), and create a package, internally it gets a new revision. It doesn't overwrite, but stored separately (you can see that in Artifactory web UI for example, or ``conan search --revisions``). This allows using previous revisions if necessary, defining them explicitly.
  The flow is very similar to the above. If you have one old, previous revision of a package in your cache, you need to ``conan install --update`` to bring the new revision to the cache and use it.

There is also an advanced topic, that is the ``package_id_mode``. Controlling it allows you to define if when a change in X is done, even if the version is not bumped, it can force a re-build, to create new binaries for its consumers.

So to summarize:

- With either revisions enabled or disabled, do a change in X without bumping the version, create the package, upload it.
- Z, and Y, need to do ``conan install --update`` to bring the new revision or changed version of X to the local cache and start using it.
- If you want to force the re-built of Z and Y when changes are done to X (and the version is not bumped), it is necessary to enable revision, and change the ``default_package_id_mode=recipe_revision_mode``, which means, the revision of the dependencies is taken into account to compute the Z and Y package-id


Please let me know if the above helps, and if you have any further question or need some pointers to read more about those topics.

**Comment by fischjo3SICKAG** [comment]
Hi @memsharded ,
thank you for the quick reply. It answers my question quite thoroughly. I understand now that  package_id_modes (e.g., recipe_revision_mode) together with lock-files should enable deterministic rebuilds of X and Z (excluding the non-determinism of the compiler).

**Comment by memsharded** [comment]
> Hi @memsharded ,
thank you for the quick reply. It answers my question quite thoroughly. I understand now that package_id_modes (e.g., recipe_revision_mode) together with lock-files should enable deterministic rebuilds of X and Z (excluding the non-determinism of the compiler).

Yes, exactly that, with ``package_id_mode`` users can select from a "relaxed" approach, that will not build new binaries, to "exact" approaches like ``recipe_revision_mode`` that guarantees one new binary to address any change in the upstream dependencies.

Lockfiles are about obtaining the exact same set of dependencies versions and revisions, even if there are new versions or revisions in the servers that could be a valid fit, so a bit more about reproducibility than deterministic builds, but you certainly want to use them, specially in CI, so new revisions uploaded to servers do not suddenly break another build. Keep an eye to Conan 1.28, because it will bring some important changes to lockfiles command line and functionality.

**Comment by fischjo3SICKAG** [comment]
I've though about saving the information from a lock file for traceability. Given that conan_build_info is currently not recommended for use, does it make sense to package the conan lock file into the package? I've tried a quick shot, but had to learn that a lock file is not always present. It is present when using 'conan install', but for 'conan create'. Is there another way (besides uploading lock file to some server)?

**Comment by memsharded** [comment]
>  Given that conan_build_info is currently not recommended for use, does it make sense to package the conan lock file into the package? I've tried a quick shot, but had to learn that a lock file is not always present. It is present when using 'conan install', but for 'conan create'. Is there another way (besides uploading lock file to some server)?

Good question. We have decided not to package the lockfile at the moment for a couple of reasons:

- Lockfiles are many times done per "application". You capture the lockfile of the whole dependency graph, and can use it to build different packages in it (like the ones affected by one upstream change). In that sense, the lockfile do not belong to just 1 package
- The lockfiles are still experimental, so they change. Next Conan 1.28 will change the format. Storing them in the packages is risky, and could render created packages unusable, and we prefer to be a bit more conservative and do not affect stability of uploaded packages.

What we are suggesting at the moment is to store them somewhere. We use a generic repository in Artifactory, and store the lockfiles (one per different configuration you create) following the naming conventions of the CI build jobs. At the user level we are still not recommending it, but suggesting that developers could resolve dependencies against a stable "develop" repo, which state would be in essence the latest successful build (and match some lockfile). 

Conan 1.28 is bringing some changes that could be interesting, for example a "versions" only base lockfile that can be used as a base for creating full lockfiles for every configuration. It is possible that some users will use it, committing it to source control, we are not sure about the use case, so we will probably learn about it from users soon. Depending on the problem you are trying to solve, we could follow-up with some discussion if you want after 1.28 is out.

**Comment by fischjo3SICKAG** [comment]
Again, thank you for your explanation. I'm looking forward to 1.28, try it out and get back to this discussion if needed.
